### ISSUE TITLE
Yii 1.1.17

### ISSUE THREAD

**Issue by allanctran** [yiisoft/yii#4002]
Updated to 1.1.17 from 1.1.16 following the upgrade instruction. Now none of my view works.

Debugging error message show that yii is unable to resolve the request, resulting in a 404 error.

**Comment by samdark** [comment]
Need more details.

**Comment by allanctran** [comment]
Hope the debug dump below helps.

array
(
    'code' => 404
    'type' => 'CHttpException'
    'errorCode' => 0
    'message' => 'Unable to resolve the request \"en/home/forms_brochures\".'
    'file' => '/var/www/framework/yiilite.php'
    'line' => 1772
    'trace' => '#0 /var/www/framework/yiilite.php(1688): CWebApplication->runController(\'en/home/forms_b...\')
#1 /var/www/framework/yiilite.php(1205): CWebApplication->processRequest()
#2 /var/www/application/index.php(145): CApplication->run()
#3 {main}'
    'traces' => array
    (
        0 => array
        (
            'file' => '/var/www/framework/yiilite.php'
            'line' => 1688
            'function' => 'runController'
            'class' => 'CWebApplication'
            'type' => '->'
            'args' => array
            (
                0 => 'en/home/forms_brochures'
            )
        )
        1 => array
        (
            'file' => '/var/www/framework/yiilite.php'
            'line' => 1205
            'function' => 'processRequest'
            'class' => 'CWebApplication'
            'type' => '->'
            'args' => array()
        )
        2 => array
        (
            'file' => '/var/www/application/index.php'
            'line' => 145
            'function' => 'run'
            'class' => 'CApplication'
            'type' => '->'
            'args' => array()
        )
    )
)

**Comment by samdark** [comment]
Nope. No idea.

**Comment by samdark** [comment]
Tried with my 1.1 projects. Works as before.

**Comment by allanctran** [comment]
I am not sure what other information to provided you with in order to debug this, the application itself run fine when on 1.1.16. But the moment that I swap that out for 1.1.17, you can see the site but none of the view file seem to load. Please advise.

**Comment by allanctran** [comment]
Ok after further digging into the yiilite.php, it seem like my route is not properly getting translate base on the rules specified in the urlManager.

Example, I have a rules like so
rules => array(
  'en/home/<caction:helloworld>' => 'mycontroller/page/<caction>'
)

when doing an echo on $route in the public function createController($route,$owner=null) method,
the route did not get translate from 'en/home/helloworld' to 'mycontroller/page/helloworld', which I think it should, because I did this same test on the version 1.1.16 and it did get translated correctly.

So I am not sure if this is a bug or some sort of enhancement.

**Comment by cebe** [comment]
related to https://github.com/yiisoft/yii/pull/3348 ?

**Comment by nkm** [comment]
Same issue here. I've just upgraded and I'm getting 404's on URIs that were working previously.

Here are some URIs:
`/pdf/de/occupancy/header`
`/pdf/de/occupancy/footer`
`/pdf/de/occupancy/welcomeSheet/19096`

And here are the matching rules:
```php
'<module:\w+>(/<lang:\w{2}>)?/<controller:\w+>/<action:\w+>/<id:\d+>' => '<module>/<controller>/<action>',
'<module:\w+>(/<lang:\w{2}>)?/<controller:\w+>/<action:\w+>' => '<module>/<controller>/<action>',
```

Any clue?

**Comment by cebe** [comment]
could someone do a git bisect to find the commit that broke it?

**Comment by lexeo** [comment]
@cebe If I'm not mistaken #3348 , added preg_quote call that escapes all special chars, so it's inpossible to use regexp in routing pattern.
I've described the problem in [this comment](https://github.com/yiisoft/yii/issues/3875#issuecomment-184873324)

**Comment by lexeo** [comment]
by my mind we encountered this problem because of using not documented features.
Optional URL params are not supported in Yii 1. I found closed issue #622
I've not found any example in the official documentation with optional params defined by Regexp.
So we can use regexps in body of param definition:
```<lang:(en|ru|de)>/fool/bar/<param1:\d+>' => 'foo/bar',```
but where did you find an example like this?:
``` '(<lang:\w{2}>/)?<controller:\w+>/<action:\w+>' => '<controller>/<action>', ````
Yes, it was working well for a long time and I also use it in my applications. As well as many [other](http://www.codedwell.com/post/78/optional-parameters-in-url-in-yii-framework) [developers](http://www.yiiframework.com/forum/index.php/topic/35616-yii-url-rules-optional-parameters/) do.
Now it doesn't work. But we have other ways to make it work:
```
 '<controller:\w+>/<action:\w+>' => '<controller>/<action>',
 '<lang:\w{2}>/<controller:\w+>/<action:\w+>' => '<controller>/<action>',
// or
[
    'route' => '<controller>/<action>',
    'pattern' => '<lang:\w{2}>/<controller:\w+>/<action:\w+>',
    'defaultParams' => ['lang' => null],
],
// or
[
    'route' => '<controller>/<action>',
    'pattern' => '(<lang:\w{2}>/)?<controller:\w+>/<action:\w+>',
    'parsingOnly' => true,
],
```
So seems we have to make some changes in our configs :-)
Please correct me if I'm wrong.

**Comment by coolcflame87** [comment]
Yii 1.1.17 + PHP 7  CHttpException not redirecting  request status to 403, 404
works fine with Yii 1.1.17 + PHP 5.6

**Comment by samdark** [comment]
@coolcflame87 looks like totally separate issue.

**Comment by allanctran** [comment]
any progress on this being verified as a bug?

**Comment by samdark** [comment]
Nope. @allanctran if you have steps to reproduce it on clean installation, it would be great.

**Comment by lexeo** [comment]
@allanctran, please show us your config section for CUrlManager component.
In debug message we can see the request URL "en/home/forms_brochures". 
So I suppose that you run into the same issue as I described above.

**Comment by allanctran** [comment]
@samdark These are the step I took on a clean installation.

    1. Install Yii 1.1.17 via zip file / upload to server
    2. created a simple webapp called 'application'
    3. created a simple module called 'helloworld'
    4. enabled the module in the config
        'modules' => array('helloworld'),
    5. added additional parameter to urlManager
        'urlFormat' => 'path',
        'urlSuffix' => '.html',
        'showScriptName' => false,
        'caseSensitive' => false,
    5. added a url route in the urlManager
        '/home/testing/<action:index>' => 'hellworld/default/<action>',
    6. added a .htaccess in webroot to remove the index.php from the url
        'RewriteEngine on'
        'RewriteCond %{REQUEST_FILENAME} !-f'
        'RewriteCond %{REQUEST_FILENAME} !-d'
        'RewriteRule . index.php'      
    7. try accessing the site https://localhost/home/testing/index.html
    8. CHttpException 404 Error - Unable to resolve the request "hellworld/default/<action>".

CHttpException output
    **/var/www/project/framework/web/CWebApplication.php(286)**

    274     public function runController($route)
    275     {
    276         if(($ca=$this->createController($route))!==null)
    277         {
    278             list($controller,$actionID)=$ca;
    279             $oldController=$this->_controller;
    280             $this->_controller=$controller;
    281             $controller->init();
    282             $controller->run($actionID);
    283             $this->_controller=$oldController;
    284         }
    285         else
    286             throw new CHttpException(404,Yii::t('yii','Unable to resolve the request "{route}".',
    287                 array('{route}'=>$route===''?$this->defaultController:$route)));
    288     }
    289 
    290     /**
    291      * Creates a controller instance based on a route.
    292      * The route should contain the controller ID and the action ID.
    293      * It may also contain additional GET variables. All these must be concatenated together   with slashes.
    294      *
    295      * This method will attempt to create a controller in the following order:
    296      * <ol>
    297      * <li>If the first segment is found in {@link controllerMap}, the corresponding
    298      * controller configuration will be used to create the controller;</li>

**main.php config file**

    'modules'=>array(
        'helloworld',
    ),

    'urlManager'=>array(
        'urlFormat'      => 'path',
        'urlSuffix'      => '.html',
        'showScriptName' => false,
        'caseSensitive'  => false,
        'rules'=>array(
            '<controller:\w+>/<id:\d+>'=>'<controller>/view',
            '<controller:\w+>/<action:\w+>/<id:\d+>'=>'<controller>/<action>',
            '<controller:\w+>/<action:\w+>'=>'<controller>/<action>',
            '/home/testing/<action:index>' => 'hellworld/default/<action>',
        ),
    ),

**Comment by lexeo** [comment]
@allanctran Hey. Did you try to debug it? You made a little mistake in config:
`
'/home/testing/<action:index>' => 'hellworld/default/<action>',
`
Just change "hellworld" to "helloworld" :)

**Comment by allanctran** [comment]
@lexeo yes I saw the error and made the correction. I've updated the post with edit notes. It seem that with Yii 1.1.17 it does not like the regex parenthesis. I've tested the route again this time using none regex route string 

    # this works.
    '/en/home/<action:index>' => 'helloworld/default/index' 

    # this does not
    '(en|fr)/home/<action:index>' => 'helloworld/default/index'

**Comment by allanctran** [comment]
OK with further investigation it is definately that Yii 1.1.17 does not like this kind of route

   # This does not - but does work in Yii 1.1.16
   # maybe this a bug in 1.1.1.6?
   '(en|fr)/home/<action:index>' => 'helloworld/default/index'

but if I do this

   # This works
   '<lang:(en|fr)>/home/<action:index>' => 'helloworld/default/index'

**Comment by lexeo** [comment]
@allanctran , yes, you are right! As I wrote in this [comment](https://github.com/yiisoft/yii/issues/3875#issuecomment-184873324) we were using the undocumented feature. Now it doesn't work and I think its normal.

**Comment by allanctran** [comment]
If this is the case how would I go about creating a route that may or may not have a parameter?
Example

     '<lang:(en|fr)>/books/(<title:\w+>/)?<page:\d>' => 'store/default/books'

Where title of the book is an optional parameter.

**Comment by samdark** [comment]
@allanctran by dividing it in two:

```php
'<lang:(en|fr)>/books/<title:\w+>/<page:\d>' => 'store/default/books'
'<lang:(en|fr)>/books/<page:\d>' => 'store/default/books'
```

**Comment by allanctran** [comment]
@samdark will this be corrected to function like 1.1.16 or remain as is going forward?

**Comment by samdark** [comment]
I think it will remain as it is now since 1.1.16 behavior wasn't documented anywhere and was not what was planned initially i.e. regex were expected to work in arguments only.
